main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -326,28 +342,288 @@ object Inliner {
   }
 
   /**
-   * returns `true` if `exp0` is considered a trivial expression.
-   *
-   * An expression is trivial if:
-   * It is either a literal (float, string, int, bool, unit), or it is a variable.
-   *
-   * A pure and trivial expression can always be inlined even without duplicating work.
+   * Recursively bind each argument in `args` to a let-expression with a fresh symbol
+   * Add corresponding symbol from `symbols` to substitution map `env0`, mapping old symbols to fresh symbols.
+   * Substitute variables in `exp0` via the filled substitution map `env0`
    */
-  private def isTrivialExp(exp0: Expression): Boolean = exp0 match {
-    case Expression.Unit(_) => true
-    case Expression.Null(_, _) => true
-    case Expression.True(_) => true
-    case Expression.False(_) => true
-    case Expression.Char(_, _) => true
-    case Expression.Float32(_, _) => true
-    case Expression.Float64(_, _) => true
-    case Expression.Int8(_, _) => true
-    case Expression.Int16(_, _) => true
-    case Expression.Int32(_, _) => true
-    case Expression.Int64(_, _) => true
-    case Expression.BigInt(_, _) => true
-    case Expression.Str(_, _) => true
-    case Expression.Var(_, _, _) => true
-    case _ => false
+  private def bindFormals(exp0: Expression, symbols: List[Symbol.VarSym], args: List[Expression], env0: Map[Symbol.VarSym, Symbol.VarSym])(implicit root: Root, flix: Flix): Expression = {
+    (symbols, args) match {
+      case (::(sym, nextSymbols), ::(e1, nextExpressions)) =>
```

Comment:
This should be `sym :: nextSymbols`, no one writes `::` as a prefix constructor.